File: lib_rxdownloader/src/main/java/com/hardy/rx/downloader/DownloadTask.java
package com.hardy.rx.downloader;

import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.IntDef;

import com.hardy.rx.downloader.model.ProgressModel;
import com.hardy.rx.downloader.model.StateModel;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.RandomAccessFile;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.concurrent.atomic.AtomicBoolean;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * @author Hdq on 2021/3/9.
 */
public class DownloadTask {
    private static final String TAG = "----DownloadTask";

    private Subject<ProgressModel> progressSubject;
    private ProgressModel progressModel = new ProgressModel();//范围 0 - 100
    private Subject<StateModel> stateSubject;
    private StateModel stateModel = new StateModel();
    private OkHttpClient client;
    private AtomicBoolean isStop = new AtomicBoolean(false);
    private String downloadUrl;
    private String localFileName;
    private float weight;//记录该任务的权重值.用于多个任务在一起时候计算总进度
    /**
     * 是否仅检查本地文件是否存在
     * true:仅检查本地文件是否存在
     * false:会通过网络检查文件是否已下载，从server拉取文件长度，检查本地文件长度是否一致，一致则认为下载完成
     * **/
    private boolean isOnlyCheckLocalFileExit = false;

    public DownloadTask(String downloadUrl, String localFileName) {
        this(downloadUrl,localFileName,1);
    }

    public DownloadTask(String downloadUrl, String localFileName, float weight) {
        if (TextUtils.isEmpty(downloadUrl)) {
            Log.e(TAG, "DownloadTask: 传入的url为空，localFileName = " + localFileName);
        }
        if (TextUtils.isEmpty(localFileName)) {
            Log.e(TAG, "DownloadTask: 传入的本地文件路径为空，downloadUrl = " + downloadUrl);
        }
        client = new OkHttpClient();
        progressSubject = PublishSubject.<ProgressModel>create().toSerialized();
        stateSubject = PublishSubject.<StateModel>create().toSerialized();
        this.downloadUrl = downloadUrl;
        this.localFileName = localFileName;

        progressModel.setTask(this);
        stateModel.setTask(this);
        stateModel.setState(State.PREPARE);
        stateSubject.onNext(stateModel);
        this.weight = weight;
    }

    /**
     * 不能直接在主线程执行，如果主线程执行，请调用{@link DownloadTask#start()}
     */
    protected void startSync(){
        isStop.getAndSet(false);
        checkDownload(downloadUrl,localFileName);
    }

    /**
     * 可在主线程直接调用
     */
    public void start(){
        Schedulers.io().scheduleDirect(() -> startSync());
    }

    public void stop(){
        isStop.getAndSet(true);
    }

    private void checkDownload(String downloadUrl, String localFileName) {
        File localFile = new File(localFileName);
        if (isOnlyCheckLocalFileExit && localFile.exists() && localFile.length() > 0) {
            notifyDownloadSuccess();
            return;
        }
        //得到下载内容的大小
        long serverContentLength = 0;
        try {
            serverContentLength = getContentLength(downloadUrl);
        } catch (IOException e) {
            progressSubject.onError(e);
            stateModel.setState(State.ERROR);
            stateSubject.onNext(stateModel);
            stateSubject.onError(e);
            return;
        }

        if(serverContentLength == 0){
            Exception e = new Exception("server返回文件长度=0");
            progressSubject.onError(e);
            stateModel.setState(State.ERROR);
            stateSubject.onNext(stateModel);
            stateSubject.onError(e);
            return;
        }
        if (localFile.exists()) {
            if (localFile.length() == serverContentLength) {//文件已经下载完
                notifyDownloadSuccess();
                return;//---!!!
            } else {//文件有错误（相当于一次错误检测）
                localFile.delete();
            }
        }
        download(serverContentLength,downloadUrl,localFileName);
    }

    private void notifyDownloadSuccess() {
        progressModel.setProgress(100);
        progressSubject.onNext(progressModel);
        progressSubject.onComplete();
        stateModel.setState(State.DOWNLOAD_SUCCESS);
        stateSubject.onNext(stateModel);
        stateSubject.onComplete();
    }

    private void download(long serverContentLength,String downloadUrl, String localFileName) {
        InputStream is = null;
        RandomAccessFile savedFile = null;
        long downloadLength = 0;   //记录已经下载的文件长度
        File tempFile = new File(localFileName + ".tmp");//临时文件
        if(tempFile.exists()){
            //如果文件存在的话，得到文件的大小
            downloadLength = tempFile.length();
            Log.d(TAG, "download: 本地临时文件大小：" + downloadLength);
        }
        Log.d(TAG, "download: server文件总大小：" + serverContentLength);
        if(serverContentLength == downloadLength){
            notifyDownloadSuccess();
            tempFile.renameTo(new File(localFileName));
            return;
        }
        /**
         * HTTP请求是有一个Header的，里面有个Range属性是定义下载区域的，它接收的值是一个区间范围，
         * 比如：Range:bytes=0-10000。这样我们就可以按照一定的规则，将一个大文件拆分为若干很小的部分，
         * 然后分批次的下载，每个小块下载完成之后，再合并到文件中；这样即使下载中断了，重新下载时，
         * 也可以通过文件的字节长度来判断下载的起始点，然后重启断点续传的过程，直到最后完成下载过程。
         */
        Request request = new Request.Builder()
                .addHeader("RANGE","bytes=" + downloadLength + "-")  //断点续传要用到的，指示下载的区间
                .url(downloadUrl)
                .build();
        try {
            Response response = client.newCall(request).execute();
            if(response != null){
                is = response.body().byteStream();
                savedFile = new RandomAccessFile(tempFile,"rw");
                savedFile.seek(downloadLength);//跳过已经下载的字节
                byte[] b = new byte[1024];
                int total = 0;
                int len;
                stateModel.setState(State.DOWNLOADING);
                stateSubject.onNext(stateModel);
                while(!isStop.get() && (len = is.read(b)) != -1){
                    total += len;
                    savedFile.write(b,0,len);
                    //计算已经下载的百分比
                    progressModel.setProgress((total + downloadLength) / (float)serverContentLength * 100f);
                    progressSubject.onNext(progressModel);
                    Log.d("----Download", "progress--0: " + progressModel.getProgress());
                }
                response.body().close();
                float progress = progressModel.getProgress();
                if (progress > 100) {
                    Log.e(TAG, "download: progress =" + progress);
                    new RuntimeException("progress =" + progress + "下载的文件大小 > server给的大小");
                }
                if (progress == 100) {
                    stateModel.setState(State.DOWNLOAD_SUCCESS);
                    boolean success = tempFile.renameTo(new File(localFileName));//重命名
                    if (!success) {
                        notifyError(new Exception("重命名失败"));
                        return;
                    }
                } else {
                    stateModel.setState(State.STOPPED);
                }
                stateSubject.onNext(stateModel);
                stateSubject.onComplete();
                progressSubject.onComplete();

            }
        } catch (IOException e) {
            e.printStackTrace();
            notifyError(e);
        } finally {
            try{
                if(is != null){
                    is.close();
                }
                if(savedFile != null){
                    savedFile.close();
                }
            }catch (Exception e){
                e.printStackTrace();
            }
        }
    }

    public void notifyError(Exception e) {
        stateModel.setState(State.ERROR);
        stateSubject.onNext(stateModel);
        stateSubject.onError(e);
        progressSubject.onError(e);
    }

    public String getDownloadUrl() {
        return downloadUrl;
    }

    public void setDownloadUrl(String downloadUrl) {
        this.downloadUrl = downloadUrl;
    }

    public String getLocalFileName() {
        return localFileName;
    }

    public void setLocalFileName(String localFileName) {
        this.localFileName = localFileName;
    }

    /**
     * 得到下载内容的大小
     * @param downloadUrl
     * @return
     */
    private long getContentLength(String downloadUrl) throws IOException {
        Request request = new Request.Builder().url(downloadUrl).build();
        Response response = client.newCall(request).execute();
        if(response !=null && response.isSuccessful()){
            long contentLength = response.body().contentLength();
            response.body().close();
            return contentLength;
        }
        return  0;
    }

    public Observable<ProgressModel> getProgressObservable() {
        return progressSubject.toSerialized();
    }

    public float getProgress() {
        return progressModel.getProgress();
    }

    public Observable<StateModel> getStateObservable() {
        return stateSubject.toSerialized();
    }

    public @State int getState() {
        return stateModel.getState();
    }

    public float getWeight() {
        return weight;
    }

    public void setWeight(float weight) {
        this.weight = weight;
    }

    public boolean isOnlyCheckLocalFileExit() {
        return isOnlyCheckLocalFileExit;
    }

    public void setOnlyCheckLocalFileExit(boolean onlyCheckLocalFileExit) {
        isOnlyCheckLocalFileExit = onlyCheckLocalFileExit;
    }

    @IntDef({State.PREPARE, State.DOWNLOADING, State.DOWNLOAD_SUCCESS, State.STOPPED, State.ERROR})
    @Retention(RetentionPolicy.SOURCE)
    public @interface State{
        int PREPARE = 0x1;
        int DOWNLOADING = 0x10;
        int DOWNLOAD_SUCCESS = 0x100;
        int STOPPED = 0x1000;

        int ERROR = -0x1;

        final class Print{
            public static String toString(@State int state){
                switch (state){
                    case PREPARE:
                        return "PREPARE";
                    case DOWNLOADING:
                        return "DOWNLOADING";
                    case DOWNLOAD_SUCCESS:
                        return "DOWNLOAD_SUCCESS";
                    case STOPPED:
                        return "STOPPED";
                    case ERROR:
                        return "ERROR";
                    default:
                        return "/";
                }
            }
        }
    }
}

File: lib_rxdownloader/src/main/java/com/wuzhu/rx/downloader/DownloadTask.kt
package com.wuzhu.rx.downloader

import android.util.Log
import androidx.annotation.Keep
import com.wuzhu.rx.downloader.exceptions.DownloadException
import com.wuzhu.rx.downloader.model.State
import com.wuzhu.rx.downloader.model.TaskProgressModel
import com.wuzhu.rx.downloader.model.TaskStateModel
import com.wuzhu.rx.downloader.utils.FilePathUtils
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import okhttp3.Call
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.File
import java.io.FileNotFoundException
import java.io.InputStream
import java.io.RandomAccessFile
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * @author Hdq on 2021/3/9.
 */
@Keep
class DownloadTask {

    companion object {
        private const val TAG = "---DownloadTask"
    }

    private val subscriptions = CompositeDisposable()
    private val progressSubject = BehaviorSubject.create<TaskProgressModel>()

    /**
     * 这里提前new出来，避免使用时候频繁创建对象
     * 范围 0 - 100
     */
    private val taskProgressModel = TaskProgressModel(this)
    private val stateSubject = BehaviorSubject.create<TaskStateModel>()

    /**
     * 这里提前new出来，避免使用时候频繁创建对象
     */
    private val taskStateModel = TaskStateModel(this)
    private var downloadCall: Call? = null
    private val isStop = AtomicBoolean(false)
    lateinit var downloadUrl: String
    lateinit var localFileName: String
    var weight: Float = 1f//记录该任务的权重值.用于多个任务在一起时候计算总进度

    /**
     * md5检测：
     * 下载前 & 下载后 检测md5
     */
    var md5: String? = null

    private val business = DownloadTaskBusiness()

    var client: OkHttpClient? = null
        get() {
            field ?: run {
                field = OkHttpClient.Builder().connectTimeout(15 * 1000, TimeUnit.MILLISECONDS).build()
            }
            return field
        }

    private val mHeaders = mutableMapOf<String, String>()

    fun create(downloadUrl: String, localFileName: String, weight: Float = 1f) {
        if (downloadUrl.isBlank()) {
            throwException("DownloadTask: 传入的url为空，localFileName = $localFileName")
        }
        if (localFileName.isBlank()) {
            throwException("DownloadTask: 传入的本地文件路径为空，downloadUrl = $downloadUrl")
        }
        if (weight <= 0) {
            throwException("DownloadTask: fileLength不能<=0")
        }
        this.downloadUrl = downloadUrl
        this.localFileName = localFileName
        this.weight = weight
        reset(true)
    }


    /**
     * 整段代码是同步执行的
     * 不能直接在主线程执行，如果主线程执行，请调用[DownloadTask.start]
     */
    fun startSync() {
        Log.w(TAG, "task--start: $this")
        try {
            isStop.getAndSet(false)
            notifyStart()
            checkDownload(downloadUrl, localFileName)
            Log.w(TAG, "task--任务执行结束: $this")
        } catch (e: Exception) {
            notifyError(e)
            Log.e(TAG, "task--任务执行错误: $this", e)
        }
    }

    /**
     * 可在主线程直接调用
     */
    fun start() {
        Schedulers.io().scheduleDirect { startSync() }
    }

    /***
     * 要支持异步调用，不能加锁，否则call?.cancel()不能立即生效
     */
    fun stop() {
        Log.e(TAG, "task--stop: $this")
        downloadCall?.let {
            if (!it.isCanceled) {
                it.cancel()
                Log.e(TAG, "task--stop--cancel--downloadCall: $this")
            }
        }
        isStop.getAndSet(true)
    }


    fun destroy() {
        stop()
        subscriptions.dispose()
    }


    private fun checkDownload(downloadUrl: String, localFileName: String) {
        if (business.checkFileNameTooLong(localFileName)) {
            throwException("文件名称过长，不能超过200个字符：localFileName=$localFileName")
        }
        md5?.let { _md5 ->
            //检测本地文件
            if (business.checkExistedLocalFile(localFileName, _md5)) {
                notifySuccess()
                return
            }
            //检测本地临时文件 - 文件已经下载完成，就差改个名字了
            val tempFileName = FilePathUtils.getTempFileName(localFileName, _md5)
            val tempFile = File(tempFileName)
            if (business.checkTempFileIsDownloadComplete(tempFile, _md5)) {
                val success = tempFile.renameTo(File(localFileName)) //重命名
                if (success) {
                    notifySuccess()
                } else {
                    throwException("temp file md5检测成功，重命名失败")
                }
                return
            }
            //检测本地临时文件 md5
            business.checkMd5BeforeDownload(tempFile, _md5)
        }
        download(downloadUrl, localFileName)
    }


    fun reset(isResetProgress: Boolean) {
        if (isResetProgress) {
            taskProgressModel.progress = 0f
        }
        progressSubject.onNext(taskProgressModel)
        notifyPrepare()
    }

    private fun download(downloadUrl: String, localFileName: String) {
        var inputStream: InputStream? = null
        var savedFile: RandomAccessFile? = null
        var downloadedLength: Long = 0 //记录已经下载的文件长度
        val tempFile = File(FilePathUtils.getTempFileName(localFileName, md5)) //临时文件
        if (tempFile.exists()) {
            //如果文件存在的话，得到文件的大小
            downloadedLength = tempFile.length()
        }
        /**
         * HTTP请求是有一个Header的，里面有个Range属性是定义下载区域的，它接收的值是一个区间范围，
         * 比如：Range:bytes=0-10000。这样我们就可以按照一定的规则，将一个大文件拆分为若干很小的部分，
         * 然后分批次的下载，每个小块下载完成之后，再合并到文件中；这样即使下载中断了，重新下载时，
         * 也可以通过文件的字节长度来判断下载的起始点，然后重启断点续传的过程，直到最后完成下载过程。
         */
        val requestBuilder = Request.Builder()
        addCustomHeaders(requestBuilder)
        val request = requestBuilder.addHeader("RANGE", "bytes=$downloadedLength-") //断点续传要用到的，指示下载的区间
            .url(downloadUrl).build()
        try {
            downloadCall = client!!.newCall(request)
            val response = downloadCall!!.execute()
            if (response.code() !in 200 until 300) {//处理错误码
                handleErrorCode(response, localFileName)
                return
            }
            response.body() ?: throwException("server返回的数据是空的")
            if (!isSupportBreakpointResume(response)) {
                Log.e(TAG, "download--: server 不支持断点;;$this")
                tempFile.deleteOnExit()
                downloadedLength = 0
            }
            response.body()?.let { responseBody ->
                val breakpointContentLength = responseBody.contentLength()
                if (isFileLengthUnKnow(breakpointContentLength)) {//文件长度未知，不支持断点
                    Log.e(TAG, "download--length: 文件长度未知，不支持断点")
                    tempFile.deleteOnExit()
                    downloadedLength = 0
                }
                val totalFileLength = downloadedLength + breakpointContentLength
                Log.e(TAG, "download--length: 已经下载的length：$downloadedLength")
                Log.e(TAG, "download--length: server返回的content-length=$breakpointContentLength")
                Log.e(TAG, "download--length: 计算的文件总大小：$totalFileLength")
                val localFile = File(localFileName)
                if (localFile.exists() && !isFileLengthUnKnow(breakpointContentLength) && localFile.length() != totalFileLength) {
                    Log.e(TAG, "download: 文件有错误,删除本地文件（一次错误检测）localFile=$localFile")
                    localFile.delete()
                }
                inputStream = responseBody.byteStream()
                //判断文件夹是否存在，不存在创建文件夹
                createTempFileDir(tempFile)
                savedFile = RandomAccessFile(tempFile, "rw")
                savedFile?.seek(downloadedLength) //跳过已经下载的字节
                val buffer = ByteArray(1024)
                var total = 0
                var len = 0
                notifyDownloading()
                while (!isStop.get() && inputStream?.read(buffer).also { len = it ?: -1 } != -1) {
                    total += len
                    savedFile?.write(buffer, 0, len)
                    //计算已经下载的百分比
                    if (isFileLengthUnKnow(breakpointContentLength)) {
                        taskProgressModel.progress = 100f
                    } else {
                        taskProgressModel.progress = (total + downloadedLength) / totalFileLength.toFloat() * 100f
                    }
                    progressSubject.onNext(taskProgressModel)
                    Log.d("----Download", "progress--0: " + taskProgressModel.progress)
                }
                response.body()?.close()
                val progress = taskProgressModel.progress
                if (progress > 100) {
                    throwException(
                        "progress =$progress,下载的文件大小 > server给的大小,localFileLength=${tempFile.length()},serverFileLength=$totalFileLength",
                        localFile,
                        tempFile,
                    )
                }
                if (progress == 100f) {
                    if (tempFile.renameTo(localFile)) {//重命名
                        if (business.checkMd5AfterDownloadComplete(localFile, md5)) {
                            notifySuccess()
                        } else {
                            throwException("下载完成，check md5 失败")
                        }
                    } else {
                        throwException("重命名失败")
                    }
                } else if (isStop.get()) {
                    Log.e(TAG, "task--stopped(没有异常): $this")
                    notifyStop(null)
                } else {
                    throwException(
                        "不是stop，且progress!=100 (说明本地文件的长度!=server文件的长度),localFileLength=${tempFile.length()},serverFileLength=$totalFileLength",
                        localFile,
                        tempFile,
                    )
                }
            }
        } catch (e: Exception) {
            if (e is FileNotFoundException) {
                Log.e(TAG, "发生严重错误！！！写本地文件失败")
            }
            handleCancelException(e)
        } finally {
            try {
                savedFile?.close()
                inputStream?.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun isSupportBreakpointResume(response: Response): Boolean {
        val result = response.code() == 206 && !response.header("Content-Range").isNullOrBlank()
        Log.d(TAG, "isSupportBreakpointResume: $result;;$this")
        return result
    }

    private fun handleErrorCode(response: Response, localFileName: String) {
        when (response.code()) {
            416 -> {//Requested Range Not Satisfiable
                val tempFile = File(FilePathUtils.getTempFileName(localFileName, md5))
                tempFile.deleteOnExit()
            }

            else -> {
            }
        }
        throwException("errorCode=${response.code()},errorBody=${response.body()?.string()}")
    }

    private fun throwException(errorMessage: String, localFile: File? = null, tempFile: File? = null) {
        localFile?.deleteOnExit()
        tempFile?.deleteOnExit()
        throw DownloadException(errorMessage)
    }

    private fun addCustomHeaders(requestBuilder: Request.Builder) {
        mHeaders.forEach {
            requestBuilder.addHeader(it.key, it.value)
        }
    }

    @Suppress("unused")
    fun addHeader(key: String, value: String) {
        mHeaders[key] = value
    }

    fun addHeaders(headers: Map<String, String>) {
        mHeaders.putAll(headers)
    }

    private fun handleCancelException(e: Exception) {
        //调用call.cancel会抛出异常
        if (isStop.get()) {
            Log.e(TAG, "task--stopped(有异常)--调用call.cancel，进入异常处理逻辑: $this", e)
            notifyStop(e)
        } else {
            notifyError(e)
        }
    }

    private fun createTempFileDir(tempFile: File) {
        tempFile.parentFile?.let {
            if (it.isFile) {
                Log.e(TAG, "发生严重错误！！！删除本地文件！！！期望是个文件夹，实际是个文件：dir=$it")
                it.deleteOnExit()
            }
            if (!it.exists()) {
                if (it.mkdirs()) {
                    Log.e(TAG, "发生严重错误！！！创建要下载的文件夹失败：dir=$it")
                }
            }
        }
    }

    private fun isFileLengthUnKnow(contentLength: Long): Boolean {
        return contentLength < 0
    }

    fun getProgressObservable(): Observable<TaskProgressModel> = progressSubject.toSerialized().doOnSubscribe {
        subscriptions.add(it)
    }.onErrorReturn {
        it.printStackTrace()
        taskProgressModel
    }

    fun getProgress(): Float = taskProgressModel.progress
    fun getStateObservable(): Observable<TaskStateModel> = stateSubject.toSerialized().doOnSubscribe {
        subscriptions.add(it)
    }.onErrorReturn {
        it.printStackTrace()
        taskStateModel.state = State.ERROR
        taskStateModel.exception = it
        taskStateModel
    }

    fun getState(): @State Int = taskStateModel.state

    override fun toString(): String {
        return "DownloadTask::${this.hashCode()}(isStop='$isStop',state='$taskStateModel',downloadUrl='$downloadUrl', localFileName='$localFileName')"
    }

    fun isTempFileExists(): Boolean {
        return File(FilePathUtils.getTempFileName(localFileName, md5)).exists()
    }

    private fun notifyPrepare() {
        taskStateModel.state = State.PREPARE
        stateSubject.onNext(taskStateModel)
        Log.w(TAG, "notifyPrepare: $this")
    }

    private fun notifyStart() {
        taskStateModel.state = State.START
        stateSubject.onNext(taskStateModel)
        Log.w(TAG, "notifyDownloading: $this")
    }

    private fun notifyDownloading() {
        taskStateModel.state = State.DOWNLOADING
        stateSubject.onNext(taskStateModel)
        Log.w(TAG, "notifyDownloading: $this")
    }

    private fun notifySuccess() {
        taskProgressModel.progress = 100f
        progressSubject.onNext(taskProgressModel)
        taskStateModel.state = State.SUCCESS
        stateSubject.onNext(taskStateModel)
        Log.w(TAG, "notifySuccess: $this")
    }

    private fun notifyStop(e: Exception?) {
        taskStateModel.state = State.STOPPED
        stateSubject.onNext(taskStateModel)
        Log.w(TAG, "notifyStop: $this", e)
    }

    fun notifyError(e: Exception?) {
        Log.w(TAG, "notifyError: $this", e)
        taskStateModel.exception = e
        taskStateModel.state = State.ERROR
        stateSubject.onNext(taskStateModel)
        isStop.getAndSet(true)
    }


}